examples/explore.py

```diff
@@ -0,0 +1,506 @@
+
+
+class Node:
+    def __init__(self, reward_sum, edge_count):
+        self.reward_sum = reward_sum
+        self.edges = [NO_EDGE] * edge_count
+        self.back_edge = None
+
+
+# Represents env states as nodes and actions as edges.
+class StateGraph:
+    def __init__(self, edges_per_node):
+        self._edges_per_node = edges_per_node
+        self._nodes = []
+        self._fingerprint_to_index = dict()
+
+    def add_or_find_node(self, fingerprint, reward_sum):
+        if fingerprint in self._fingerprint_to_index:
+            node_index = self._fingerprint_to_index[fingerprint]
+            assert rewards_close(self._nodes[node_index].reward_sum, reward_sum)
+            return (node_index, False)
+        node_index = self.node_count()
+        self._fingerprint_to_index[fingerprint] = node_index
+        node = Node(reward_sum, self._edges_per_node)
+        self._nodes.append(node)
+        return (node_index, True)
+
+    def add_edge(self, from_node_index, edge_index, to_node_index):
+        assert edge_index in range(self._edges_per_node)
+        assert from_node_index in range(self.node_count())
+        assert to_node_index in range(self.node_count())
+        assert self.get_edge(from_node_index, edge_index) == NO_EDGE
+
+        from_node = self._nodes[from_node_index]
+        from_node.edges[edge_index] = to_node_index
+
+        to_node = self._nodes[to_node_index]
+        if to_node.back_edge is None:
+            to_node.back_edge = (from_node_index, edge_index)
+
+    def get_edge(self, from_node_index, edge_index):
+        assert edge_index < self._edges_per_node
+        assert from_node_index < self.node_count()
+        return self._nodes[from_node_index].edges[edge_index]
+
+    # Returns a path back to node 0. For this to work, edges have to
+    # be added in a order so that the subgraph consisting of the first
+    # in-coming edge to each node defines a tree with node 0 as the
+    # root.
+    def node_path(self, node_index):
+        assert node_index < self.node_count()
+
+        path = []
+        while node_index != 0:
+            back_edge = self._nodes[node_index].back_edge
+            assert back_edge is not None
+            (prior_node_index, edge_index) = back_edge
+            node_index = prior_node_index
+            path.append(edge_index)
+        path.reverse()
+        return path
+
+    def reward_sum(self, node_index):
+        return self._nodes[node_index].reward_sum
+
+    def node_count(self):
+        return len(self._nodes)
+
+
+def env_to_fingerprint(env):
+    # TODO: There is some sort of state in the env that is not
+    # captured by this. Figure out what it is and fix it. Also
+    # consider adding a fingerprint observation to env.
+    if False:
+        # BitcodeFile is slower, so using Ir instead.
+        path = env.observation["BitcodeFile"]
+        with open(path, "rb") as f:
+            data = f.read()
+    else:
+        data = env.observation["Ir"].encode()
```

Comment:
```suggestion
    data = env.observation["Ir"].encode()
```

The `BitcodeFile` and `Ir` observations call [different LLVM APIs](https://github.com/facebookresearch/CompilerGym/blob/development/compiler_gym/envs/llvm/service/LlvmEnvironment.cc#L209-L230), but I think the approach you used is exactly right for what you want.

Something to keep in mind for later would be to add an observation space that returns the [Benchmark::hash()](https://github.com/facebookresearch/CompilerGym/blob/development/compiler_gym/envs/llvm/service/Benchmark.h#L69) SHA1, rather than serializing the entire IR and doing the checksum from the python side. Not required for this PR.

Reply:
I think it makes sense leaving the TODO here indicating both that something is not captured, so that this isn't a real fingerprint even though the function is named as though it is (which otherwise could trip someone up the same way it did me for a while), and that having this on the env itself as an observation is preferable. I'd also like to keep both IR/bitcode versions of computing the "fingerprint" since (surprisingly) they can distinguish separate sets of envs and e.g. one thing that would be logical to do in adding a fingerprint observation on the env is to determine whether the set of envs distinguished by the bitcode fingerprint here is identical to that distinguished by the fingerprint on the env. Which it ought to be, but I'm not confident, e.g. I wonder what the interplay here is with the option "ShouldPreserveUseListOrder" used when calculating the hash / bitcode file and any other options. Once this is all cleared up then the entire function should be removed.

Reply:
Sounds good, thanks for the explanation.